File: app/src/main/java/com/njlabs/showjava/utils/ExceptionHandler.java
package com.njlabs.showjava.utils;

import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.widget.Toast;

import com.crashlytics.android.Crashlytics;
import com.njlabs.showjava.ui.JavaExplorer;

import java.io.PrintWriter;
import java.io.StringWriter;

/**
 * Created by Niranjan on 27-05-2015.
 */
@SuppressWarnings({"FieldCanBeLocal", "StringBufferReplaceableByString"})
public class ExceptionHandler implements java.lang.Thread.UncaughtExceptionHandler {

    private final Context myContext;
    private final String LINE_SEPARATOR = "\n";
    private final String sourceDir;
    private final String packageID;

    public ExceptionHandler(Context context, String sourceDir, String packageID) {
        this.myContext = context;
        this.sourceDir = sourceDir;
        this.packageID = packageID;
    }

    public void uncaughtException(Thread thread, Throwable exception) {

        Crashlytics.logException(exception);

        StringWriter stackTrace = new StringWriter();
        exception.printStackTrace(new PrintWriter(stackTrace));
        StringBuilder errorReport = new StringBuilder();
        errorReport.append("************ CAUSE OF ERROR ************\n\n");
        errorReport.append(stackTrace.toString());

        Log.e("com.njlabs.showjava", errorReport.toString());

        Toast.makeText(myContext, "There was an error decompiling this app. Showing incomplete source.", Toast.LENGTH_LONG).show();
        Intent intent = new Intent(myContext, JavaExplorer.class);
        intent.putExtra("java_source_dir", sourceDir);
        intent.putExtra("package_id", packageID);
        myContext.startActivity(intent);

        android.os.Process.killProcess(android.os.Process.myPid());
        System.exit(10);
    }

}

File: app/src/main/kotlin/com/njlabs/showjava/utils/ExceptionHandler.kt
/*
 * Show Java - A java/apk decompiler for android
 * Copyright (c) 2018 Niranjan Rajendran
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.njlabs.showjava.utils

import android.content.Context
import android.content.Intent
import android.widget.Toast
import com.njlabs.showjava.activities.explorer.navigator.NavigatorActivity
import timber.log.Timber
import java.io.PrintWriter
import java.io.StringWriter


class ExceptionHandler(
    private val context: Context,
    private val sourceDir: String,
    private val packageID: String
) : java.lang.Thread.UncaughtExceptionHandler {

    private val LINE_SEPARATOR = "\n"

    override fun uncaughtException(thread: Thread, exception: Throwable) {
        val stackTrace = StringWriter()
        exception.printStackTrace(PrintWriter(stackTrace))
        val errorReport = StringBuilder()
        errorReport.append("************ CAUSE OF ERROR ************\n\n")
        errorReport.append(stackTrace.toString())

        Timber.e(exception)

        Toast.makeText(
            context,
            "There was an error decompiling this app. Showing incomplete source.",
            Toast.LENGTH_LONG
        ).show()

        val intent = Intent(context, NavigatorActivity::class.java)
        intent.putExtra("java_source_dir", sourceDir)
        intent.putExtra("package_id", packageID)
        context.startActivity(intent)

        android.os.Process.killProcess(android.os.Process.myPid())
        System.exit(10)
    }

}
